staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -205,57 +398,59 @@ func (q *delayingType) waitingLoop() {
 			return
 		}
 
-		now := q.clock.Now()
-
 		// Add ready entries
+		popped := false
 		for waitingForQueue.Len() > 0 {
 			entry := waitingForQueue.Peek().(*waitFor)
-			if entry.readyAt.After(now) {
+			if entry.readyAt.After(q.clock.Now()) {
 				break
 			}
-
+			if q.logger != nil {
+				q.log(fmt.Sprintf("waitingLoop: item matured from waiting: %s", entry))
+			}
 			entry = heap.Pop(waitingForQueue).(*waitFor)
-			q.Add(entry.data)
+			popped = true
+			q.Interface.Add(entry.data)
 			delete(waitingEntryByData, entry.data)
 		}
-
-		// Set up a wait for the first item's readyAt (if one exists)
-		nextReadyAt := never
-		if waitingForQueue.Len() > 0 {
-			if nextReadyAtTimer != nil {
-				nextReadyAtTimer.Stop()
-			}
-			entry := waitingForQueue.Peek().(*waitFor)
-			nextReadyAtTimer = q.clock.NewTimer(entry.readyAt.Sub(now))
-			nextReadyAt = nextReadyAtTimer.C()
+		// recalculate the nextReadyAtTimer when we pop items out of the queue.
+		if popped {
+			q.calculateNextReadyAt(waitingForQueue)
 		}
 
 		select {
 		case <-q.stopCh:
+			if q.logger != nil {
+				q.log("waitingLoop: received STOP signal")
+			}
 			return
 
 		case <-q.heartbeat.C():
+			if q.logger != nil {
+				q.log("waitingLoop: received HEARTBEAT event")
+			}
 			// continue the loop, which will add ready items
 
-		case <-nextReadyAt:
+		case <-q.nextReadyAtTimerCh:
+			if q.logger != nil {
+				q.log("waitingLoop: received NEXT-READY-AT event")
+			}
 			// continue the loop, which will add ready items
 
 		case waitEntry := <-q.waitingForAddCh:
-			if waitEntry.readyAt.After(q.clock.Now()) {
-				insert(waitingForQueue, waitingEntryByData, waitEntry)
-			} else {
-				q.Add(waitEntry.data)
+			if q.logger != nil {
+				q.log("waitingLoop: received new item: " + waitEntry.String())
 			}
+			q.processArrivingEntry(waitingForQueue, waitingEntryByData, waitEntry)
 
 			drained := false
 			for !drained {
 				select {
 				case waitEntry := <-q.waitingForAddCh:
-					if waitEntry.readyAt.After(q.clock.Now()) {
-						insert(waitingForQueue, waitingEntryByData, waitEntry)
-					} else {
-						q.Add(waitEntry.data)
+					if q.logger != nil {
+						q.log("waitingLoop: receiving additional item: " + waitEntry.String())
```

Comment:
No, I wanted to know if the item triggered the waitLoop or was added by the different code path which checks for other items having become ready, I don't want them to look the same.

Reply:
I suppose, from our debugging perspective. But it has the same meaning for the end user, in case we want to expose this as an interface. Meaning they would probably not care about this distinction , they just want to know their items are getting drained

Reply:
The logging is only used for debugging and then it is useful to be able to distinguish between an add that triggered the waitingForAddCh event or piggy-backed through behind another item.